internal/devserver/devserver.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/bazelbuild/rules_typescript/internal/concatjs"
+)
+
+var (
+	port        = flag.Int("port", 5432, "server port to listen on")
+	base        = flag.String("base", "", "server base (required, runfiles of the binary)")
+	pkgs        = flag.String("packages", "", "root package(s) to serve, comma-separated")
+	manifest    = flag.String("manifest", "", "sources manifest (.MF)")
+	servingPath = flag.String("serving_path", "/_/ts_scripts.js", "path to serve the combined sources at")
+)
+
+type customNotFoundResponseWriter struct {
+	http.ResponseWriter
+
+	request  *http.Request
+	notFound http.HandlerFunc
+	has404   bool
+	hasWrite bool
+}
+
+// Write implements http.ResponseWriter.Write.
+func (w *customNotFoundResponseWriter) Write(b []byte) (int, error) {
+	w.hasWrite = true
+	if w.has404 {
+		// We have already written the not found response, so drop this one.
+		return len(b), nil
+	}
+	return w.ResponseWriter.Write(b)
+}
+
+// WriteHeader implements http.ResponseWriter.WriteHeader.
+func (w *customNotFoundResponseWriter) WriteHeader(code int) {
+	if code != http.StatusNotFound || w.hasWrite {
+		// We only intercept not found statuses. We also don't intercept statuses written after the
+		// first write as these are an error and should be handled by the default ResponseWriter.
+		w.ResponseWriter.WriteHeader(code)
+		return
+	}
+
+	// WriteHeader writes out the entire header (including content type) and only the first call
+	// will succeed. Therefore, if we want the correct content type set, we must set it here.
+	w.Header().Del("Content-Type")
+	w.Header().Add("Content-Type", "text/html; charset=utf-8")
+	w.ResponseWriter.WriteHeader(code)
+	w.has404 = true
+
+	// We have already written the header, so drop any calls to WriteHeader made by the not found
+	// handler. These additional calls are expected, and if passed through, would cause the base
+	// ResponseWriter to unnecessarily spam the error log.
+	w.notFound(&headerSuppressorResponseWriter{w.ResponseWriter}, w.request)
+	w.hasWrite = true
+}
+
+type headerSuppressorResponseWriter struct {
+	http.ResponseWriter
+}
+
+// WriteHeader implements http.ResponseWriter.WriteHeader.
+func (w *headerSuppressorResponseWriter) WriteHeader(code int) {}
+
+func createFileHandler(servingPath, manifest string, pkgs []string, base string) http.HandlerFunc {
+	pkgPaths := chainedDir{}
```

Comment:
In concatjs these are called 'roots', maybe we should stick to that?  "pkg" doesn't really convey that they are paths to resolve from